File: adapters/telemetry/tests/test_quantizer.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 Crystal Warden Supply Chain Labs LLC
"""Decision-preserving quantizer — the contract, pinned:
  * the partition is the coarsest that keeps every atom truth-invariant per cell (minimum sufficient
    statistic for the decisions);
  * quantize -> reconstruct routes IDENTICALLY to the original reading, across numeric / boolean /
    categorical fields (the differentiated claim, in miniature — the frozen cross-flow proof is next);
  * symbols are small ints (Fibonacci-friendly).
"""
import os
import sys
from pathlib import Path

import pytest

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))
import quantizer as q  # noqa: E402

from prismpath import predicates  # noqa: E402
from prismpath.parser import parse, parse_file  # noqa: E402

_INCIDENT = os.path.join(os.path.dirname(__file__), "..", "..", "..",
                         "prismpath", "gallery", "incident_severity", "incident_severity.md")

CATEGORICAL = """---
name: cat
start: classify
---
## classify
-> urgent: when kind == 'urgent'
-> batch: when kind in ('nightly', 'weekly')
-> blocked: when status != 'ok'
-> normal: else
## urgent
## batch
## blocked
## normal
"""

NUMERIC_EQ = """---
name: numeq
start: classify
---
## classify
-> exact: when x == 5
-> high: when x >= 10
-> low: else
## exact
## high
## low
"""


def _route(graph, node, reading):
    """First-match deterministic routing from `node` over a reading (mirrors the engine's edge order)."""
    for target, cond in graph.nodes[node].edges:
        if predicates.is_deterministic(cond) and predicates.eval_condition(cond, reading):
            return target
    return None


def _assert_decisions_preserved(graph, node, readings):
    parts = q.build_partitions(graph)
    for r in readings:
        orig = _route(graph, node, r)
        recon = _route(graph, node, q.reconstruct(parts, q.quantize(parts, r)))
        assert orig == recon, f"decision changed for {r}: {orig} -> {recon}"


# ---------------------------------------------------------------- partition shape
def test_incident_partition_is_minimal():
    g = parse_file(_INCIDENT)
    parts = q.build_partitions(g)
    # error_rate thresholds >=1,>=5,>=25 -> exactly 4 decision cells
    assert parts["error_rate"].kind == "numeric"
    assert parts["error_rate"].n == 4
    # the two bare-field flags are boolean (2 cells each)
    assert parts["data_at_risk"].kind == "boolean" and parts["data_at_risk"].n == 2
    assert parts["user_facing"].kind == "boolean" and parts["user_facing"].n == 2


def test_categorical_partition_shape():
    g = parse(CATEGORICAL)
    parts = q.build_partitions(g)
    # kind: urgent/nightly/weekly + other = 4 ; status: ok + other = 2
    assert parts["kind"].kind == "categorical" and parts["kind"].n == 4
    assert parts["status"].kind == "categorical" and parts["status"].n == 2


def test_numeric_equality_keeps_the_point_cell():
    g = parse(NUMERIC_EQ)
    parts = q.build_partitions(g)
    # x: {<5}, {5}, {6..9}, {>=10} -> 4 cells (the == carves out the singleton {5})
    assert parts["x"].kind == "numeric" and parts["x"].n == 4
    assert parts["x"].symbol(5) != parts["x"].symbol(4)
    assert parts["x"].symbol(7) != parts["x"].symbol(5)


# ---------------------------------------------------------------- the focused decision-preserving proof
def test_incident_decisions_preserved():
    g = parse_file(_INCIDENT)
    readings = []
    for dar in (True, False):
        for uf in (True, False):
            for er in [-5, 0, 1, 2, 4, 5, 6, 24, 25, 26, 50, 100]:
                readings.append({"data_at_risk": dar, "user_facing": uf, "error_rate": er})
    _assert_decisions_preserved(g, "classify" if "classify" in g.nodes else "assess", readings)


def test_categorical_decisions_preserved():
    g = parse(CATEGORICAL)
    readings = [{"kind": k, "status": s}
                for k in ("urgent", "nightly", "weekly", "adhoc", "xyz")
                for s in ("ok", "bad", "degraded")]
    _assert_decisions_preserved(g, "classify", readings)


def test_numeric_equality_decisions_preserved():
    g = parse(NUMERIC_EQ)
    _assert_decisions_preserved(g, "classify", [{"x": v} for v in range(-3, 20)])


# ---------------------------------------------------------------- symbols stay small
def test_symbols_are_small():
    g = parse_file(_INCIDENT)
    parts = q.build_partitions(g)
    r = {"data_at_risk": True, "user_facing": False, "error_rate": 42}
    syms = q.quantize(parts, r)
    assert all(0 <= s < 8 for s in syms.values())   # tiny -> 1-2 Fibonacci bytes each

File: prismpath-telemetry-rs/tests/test_quantizer.rs
use prismpath_rs::{parse, V};
use prismpath_telemetry_rs::quantizer;
use std::collections::HashMap;

const INCIDENT: &str = r#"---
name: incident_severity
start: assess
---

## assess
Read the incoming alert and classify it. Emit `user_facing` (bool), `error_rate` (percent of
requests failing, 0–100), and `data_at_risk` (bool).
@emits(user_facing, error_rate, data_at_risk)
-> sev1_page: when data_at_risk
-> sev1_page: when user_facing and error_rate >= 25
-> sev2_oncall: when user_facing and error_rate >= 5
-> sev3_ticket: when error_rate >= 1
-> watch: else

## sev1_page
## sev2_oncall
## sev3_ticket
## watch
"#;

const CATEGORICAL: &str = r#"---
name: cat
start: classify
---
## classify
-> urgent: when kind == 'urgent'
-> batch: when kind in ('nightly', 'weekly')
-> blocked: when status != 'ok'
-> normal: else
## urgent
## batch
## blocked
## normal
"#;

const NUMERIC_EQ: &str = r#"---
name: numeq
start: classify
---
## classify
-> exact: when x == 5
-> high: when x >= 10
-> low: else
## exact
## high
## low
"#;

fn route(graph: &prismpath_rs::Graph, node: &str, reading: &HashMap<String, V>) -> Option<String> {
    if let Some(flow_node) = graph.nodes.get(node) {
        for (target, cond) in &flow_node.edges {
            if prismpath_rs::is_deterministic(cond) {
                if let Ok(true) = prismpath_rs::eval_condition(cond, reading) {
                    return Some(target.clone());
                }
            }
        }
    }
    None
}

fn assert_decisions_preserved(graph: &prismpath_rs::Graph, node: &str, readings: &[HashMap<String, V>]) {
    let parts = quantizer::build_partitions(graph);
    for reading in readings {
        let orig = route(graph, node, reading);
        let recon = route(graph, node, &quantizer::reconstruct(&parts, &quantizer::quantize(&parts, reading).unwrap()));
        assert_eq!(orig, recon, "decision changed for {:?}", reading);
    }
}

#[test]
fn test_incident_partition_is_minimal() {
    let graph = parse(INCIDENT);
    let parts = quantizer::build_partitions(&graph);
    assert_eq!(parts["error_rate"].kind, quantizer::FieldKind::Numeric);
    assert_eq!(parts["error_rate"].n, 4);
    assert_eq!(parts["data_at_risk"].kind, quantizer::FieldKind::Boolean);
    assert_eq!(parts["data_at_risk"].n, 2);
    assert_eq!(parts["user_facing"].kind, quantizer::FieldKind::Boolean);
    assert_eq!(parts["user_facing"].n, 2);
}

#[test]
fn test_categorical_partition_shape() {
    let graph = parse(CATEGORICAL);
    let parts = quantizer::build_partitions(&graph);
    assert_eq!(parts["kind"].kind, quantizer::FieldKind::Categorical);
    assert_eq!(parts["kind"].n, 4);
    assert_eq!(parts["status"].kind, quantizer::FieldKind::Categorical);
    assert_eq!(parts["status"].n, 2);
}

#[test]
fn test_numeric_equality_keeps_the_point_cell() {
    let graph = parse(NUMERIC_EQ);
    let parts = quantizer::build_partitions(&graph);
    assert_eq!(parts["x"].kind, quantizer::FieldKind::Numeric);
    assert_eq!(parts["x"].n, 4);
    assert_ne!(parts["x"].symbol(&V::Num(5.0)).unwrap(), parts["x"].symbol(&V::Num(4.0)).unwrap());
    assert_ne!(parts["x"].symbol(&V::Num(7.0)).unwrap(), parts["x"].symbol(&V::Num(5.0)).unwrap());
}

#[test]
fn test_incident_decisions_preserved() {
    let graph = parse(INCIDENT);
    let mut readings = Vec::new();
    for dar in [true, false] {
        for uf in [true, false] {
            for er in [-5, 0, 1, 2, 4, 5, 6, 24, 25, 26, 50, 100] {
                let mut reading = HashMap::new();
                reading.insert("data_at_risk".to_string(), V::Bool(dar));
                reading.insert("user_facing".to_string(), V::Bool(uf));
                reading.insert("error_rate".to_string(), V::Num(er as f64));
                readings.push(reading);
            }
        }
    }
    assert_decisions_preserved(&graph, "assess", &readings);
}

#[test]
fn test_categorical_decisions_preserved() {
    let graph = parse(CATEGORICAL);
    let mut readings = Vec::new();
    for kind in ["urgent", "nightly", "weekly", "adhoc", "xyz"] {
        for status in ["ok", "bad", "degraded"] {
            let mut reading = HashMap::new();
            reading.insert("kind".to_string(), V::Str(kind.to_string()));
            reading.insert("status".to_string(), V::Str(status.to_string()));
            readings.push(reading);
        }
    }
    assert_decisions_preserved(&graph, "classify", &readings);
}

#[test]
fn test_numeric_equality_decisions_preserved() {
    let graph = parse(NUMERIC_EQ);
    let readings: Vec<HashMap<String, V>> = (-3..20)
        .map(|value| {
            let mut reading = HashMap::new();
            reading.insert("x".to_string(), V::Num(value as f64));
            reading
        })
        .collect();
    assert_decisions_preserved(&graph, "classify", &readings);
}

#[test]
fn test_symbols_are_small() {
    let graph = parse(INCIDENT);
    let parts = quantizer::build_partitions(&graph);
    let mut reading = HashMap::new();
    reading.insert("data_at_risk".to_string(), V::Bool(true));
    reading.insert("user_facing".to_string(), V::Bool(false));
    reading.insert("error_rate".to_string(), V::Num(42.0));
    let syms = quantizer::quantize(&parts, &reading).unwrap();
    for symbol in syms.values() {
        assert!(*symbol < 8);
    }
}

#[test]
fn test_out_of_partition_numeric_is_err_not_panic() {
    // build_partitions always leaves open ends, so construct a bounded partition by hand — the
    // case a library consumer (e.g. a Vector codec) could feed; it must be an Err, never a panic.
    let cells = vec![quantizer::Cell { lo: Some(0), hi: Some(9), const_val: None, rep: V::Num(5.0) }];
    let part = quantizer::FieldPartition::new("x".into(), quantizer::FieldKind::Numeric, cells);
    assert!(part.symbol(&V::Num(5.0)).is_ok());
    assert!(part.symbol(&V::Num(99.0)).is_err());
    let mut parts = HashMap::new();
    parts.insert("x".to_string(), part);
    let mut reading = HashMap::new();
    reading.insert("x".to_string(), V::Num(99.0));
    assert!(quantizer::quantize(&parts, &reading).is_err());
}
